### ISSUE TITLE
Pretrain data and GPU specs

### ISSUE THREAD

**Issue by ngkayjay** [zcao0420/MOFormer#3]
Hi! I'm Marcus Ng, and we have corresponded previously. You mentioned that the original dataset of ~400K cif files and mofids used to joint pre-train the models together are around 4GB and uploading it is not tractable (I would say there are many ways outside of github which offer the opportunity to open source just 4GB data, but I acknowledge you have both graduated and moved on), and have instead provided us with 100 toy cif files. 
That being said, I need to verify your results. I am attempting to reproduce your dataset from CORE-MOF-2019, hMOF and Boyd & Woo from scratch. If I follow your data (id_prop file with [[<cif file string>, <mof id string>], ...] but 400k len instead of 100) and directory formats, will running the pretrain_SSL.py allow me to properly pre-train the models to the same specification as your paper?
Thanks!

**Comment by zcao0420** [comment]
Hello Marcus,

We are currently trying to find an online data repo to host the >400K pretraining data. I will leave this issue open while we are doing that and inform you once the data is ready. I believe that pretrain_SSL.py should allow you to pretrain the models given the data. You can follow up with us if you have any trouble running the pretrain_SSL.py file.

**Comment by RishikeshMagar** [comment]
Hi @ngkayjay ,
Did you get the toy file working?

**Comment by RishikeshMagar** [comment]
@ngkayjay I think this should work, we uploaded the data last week. 

https://figshare.com/articles/journal_contribution/cif_tar_xz/23532918

Initially, I thought there was 24 hour latency for fig share and that is why the link did not work. However, the way fig share works is slightly different and link was intractable. We will update the protocol for data download and hopefully it will be smooth sailing thereafter. 

Running `pretrain_SSL.py` will allow you to properly pre-train the models as specified by the paper. The pretraining was done on NVIDIA RTX3090 and it required us around 5 days to pretrain for 15 epochs. The GPU ram was 24 GB and the training was done on a single GPU. Of course depending on the power of your GPU you may observe different statistics.

**Comment by ngkayjay** [comment]
Thank you for your data, advice, and continued support!
I've just managed to get the data pre-processing pipeline working on my side as well, to generate mofids for an arbitrary set of .cif files, as input for joint pretraining of the MOFormer and CGCNN.

I'd like to ask, in order to prepare your dataset (Boyd&Woo+CoRE2019+hMOF), how long did it take for you to construct the mofids for:
-one .cif file, and
-all 400K+ .cif files?
I assume you used the python package as implemented by the Snurr group (https://github.com/snurr-group/mofid/tree/master)?

**Comment by zcao0420** [comment]
Yes, I used the package from Snurr Group to convert cif to mofid. I modified their code a little bit so it outputs only a string (mofid) for each cif, instead of a bunch of other unnecessary things to our project. Each conversion takes about 6-7 seconds on my local desktop but varies a lot by the size of the MOF. I would highly suggest multithreading because otherwise it can take >1week to convert >400K cif files.

**Comment by zcao0420** [comment]
@ngkayjay We have fixed some bugs for the past few days. There was a version mismatch between local code files and the Github ones. Hopefully it should be good now. The link to figshare is also fixed, so you can download all the cif files and their corresponding MOFid for all >400K pretraining data. Please let us know if there is any further issue.

**Comment by ngkayjay** [comment]
@zcao0420 Thanks!

For setup: While I didn't experience an error on colab, I had to use tensorboardx instead of tensorboard on my HPC. Might just be something to do with versioning.

After env setup, running `python pretrain_SSL.py`, it appears that the code is looking for a `atom_init.json` in the directory containing the cif files and `id_prop.npy`. There isn't one in your cif_toy folder, and there isn't one in the cif.tar. I copied the one from your benchmark_datasets on this repo. I assume this is the correct file?

**Comment by zcao0420** [comment]
Yes, that's it.  Thanks for catching it.

**Comment by ngkayjay** [comment]
Ok, it seems to work now. In addition to copying the file over, having the `atom_init.json` file in the cif directory (as stated in the docs for CIFData) would require a change to 'line 360' of `dataset/dataset_multiview.py` from
360 | `atom_init_file = os.path.join('dataset/atom_init.json')`
to
360 | `atom_init_file = os.path.join(self.root_dir, 'atom_init.json')`.

**Comment by ngkayjay** [comment]
Running pretrain_SSL.py on the fixed cif_toy, with epochs set to 20 in config_multiview, I get:
`
Pre-trained weights not found. Training from scratch.
0 0 562.5459594726562
Validation 554.4138793945312
Validation 575.8638305664062
Validation 549.0695190429688
Validation 590.9860229492188
Validation 551.85546875
Validation 561.5661010742188
Validation 552.1605224609375
Validation 550.1588745117188
Validation 550.7239379882812
Validation 541.9981689453125
Validation 555.5213623046875
Validation 538.6709594726562
Validation 549.1261596679688
Validation 540.9842529296875
Validation 542.5035400390625
Validation 542.3846435546875
Validation 526.9327392578125
Validation 549.7733154296875
Validation 533.5387573242188
Validation 529.5744018554688
`
It only seems to print any Barlow Loss for the first epoch, and then afterwards only prints validation loss. Is the joint SSL training working as intended?

**Comment by RishikeshMagar** [comment]
```
 if n_iter % self.config['log_every_n_steps'] == 0:
                    self.writer.add_scalar('train_loss', loss.item(), global_step=n_iter)
                    self.writer.add_scalar('cosine_lr_decay', scheduler.get_last_lr()[0], global_step=n_iter)
                    print(epoch_counter, bn, loss.item())
```

You will have to make changes to this part of the code to print losses after every epoch. This paper has some legacy codes from some of the previous SSL works done in the lab such as MolCLR and CT.

**Comment by ngkayjay** [comment]
Authors, it seems like we've managed to get the joint-pretraining going! We report 1 epoch of 413k cif files (full datast) takes ~8 hours on a A100 with 16 cpu cores, in line with what you've reported (weird that there's no speedup though).
We report a training loss of ~113 units and validation loss of ~117 after the first epoch.
However, we observed that when running the code from the github as is, validation seems to take a whole 8 hours, which has to be unintended behavior. Digging in the code it appears that validation ratio is default ta 0.1, so I'm wondering if I'm missing out anything here.
Thanks for your continued support!

**Comment by RishikeshMagar** [comment]
Are you using the config_multiview.yaml?The val ratio is set via that file, the file in the repo sets val ratio to 0.5. Since, your train and val ratio is the same, it takes the same time for both training and validation. Set the val ratio to 0.05 and hopefully it will be resolved. This is my guess of what might be happening, I can't be 100% sure, do let us know if that is not the case 

The val ratio is 0.5 on GitHub in accordance to the toy data(100 files), if we had set 0.05 it would have been negligible files in the validation set so I think I set to 0.5 when testing everything locally.

**Comment by ngkayjay** [comment]
Oh I see, thanks! What batch size did you guys use for your joint pre-training? Also, approx. what's your training and validation losses (Barlow Twins) are you getting after 15 epochs?

**Comment by RishikeshMagar** [comment]
I don't remember the loss values off the top of my head maybe I will check and get back to you later if other issues arise. Regarding the batch size and other hyperparameters you should refer to the SI of the paper. I am closing this issue now since you got the framework up and running. Feel free to reopen if any other issues arise. Good luck for your research. 

One thing we could not investigate but you may try is examine diminishing returns of pretraining, in terms of epochs. Are just a few training epochs sufficient can more training epochs benefit during fine-tuning.
